src/js/libcs/Render2D.js

```diff
@@ -500,59 +500,41 @@ Render2D.drawpoint = function(pt) {
 };
 
 Render2D.drawline = function(homog) {
-    var na = CSNumber.abs(homog.value[0]).value.real;
-    var nb = CSNumber.abs(homog.value[1]).value.real;
-    var nc = CSNumber.abs(homog.value[2]).value.real;
-    var divi;
-
-    if (na >= nb && na >= nc) {
-        divi = homog.value[0];
-    }
-    if (nb >= na && nb >= nc) {
-        divi = homog.value[1];
+    if (!List._helper.isAlmostReal(homog))
+        return;
+    var l = List.normalizeMax(homog);
+    l = [l.value[0].value.real, l.value[1].value.real, l.value[2].value.real];
+
+    function crossxy(u, v) {
+        var z = (u[0] * v[1] - u[1] * v[0]);
+        return {
+            x: (u[1] * v[2] - u[2] * v[1]) / z,
+            y: (u[2] * v[0] - u[0] * v[2]) / z
+        };
     }
-    if (nc >= nb && nc >= na) {
-        divi = homog.value[2];
+
+    function sortxy(u, v) {
+        if (u.x !== v.x)
+            return u.x - v.x;
+        else
+            return u.y - v.y;
     }
-    var a = CSNumber.div(homog.value[0], divi);
-    var b = CSNumber.div(homog.value[1], divi);
-    var c = CSNumber.div(homog.value[2], divi); //TODO Realitycheck einbauen
-
-    var l = [
-        a.value.real,
-        b.value.real,
-        c.value.real
-    ];
-    var b1, b2;
-    if (Math.abs(l[0]) < Math.abs(l[1])) {
-        b1 = [1, 0, 30];
-        b2 = [-1, 0, 30];
-    } else {
-        b1 = [0, 1, 30];
-        b2 = [0, -1, 30];
+
+    // do not draw beyond canvas boundary
+    var m = csport.drawingstate.matrix;
+    var b1 = [m.a, m.b, m.tx]; // left boundary
+    var b2 = [m.a, m.b, m.tx - csctx.canvas.width]; // right boundary
+    var b3 = [m.c, m.d, m.ty]; // upper boundary
+    var b4 = [m.c, m.d, csctx.canvas.height - m.ty]; // lower boundary
+    if (l[0] === 0) { // horizontal line
+        Render2D.drawsegcore(crossxy(l, b1), crossxy(l, b2));
+        return;
+    } else if (l[1] === 0) { // vertical line
+        Render2D.drawsegcore(crossxy(l, b3), crossxy(l, b4));
+        return;
     }
-    var erg1 = [
-        l[1] * b1[2] - l[2] * b1[1],
-        l[2] * b1[0] - l[0] * b1[2],
-        l[0] * b1[1] - l[1] * b1[0]
-    ];
-    var erg2 = [
-        l[1] * b2[2] - l[2] * b2[1],
-        l[2] * b2[0] - l[0] * b2[2],
-        l[0] * b2[1] - l[1] * b2[0]
-    ];
-
-    var pt1 = {
-        x: erg1[0] / erg1[2],
-        y: erg1[1] / erg1[2]
-    };
-    var pt2 = {
-        x: erg2[0] / erg2[2],
-        y: erg2[1] / erg2[2]
-
-    };
-
-    Render2D.drawsegcore(pt1, pt2);
+    var intersection = [b1, b2, b3, b4].map(crossxy.bind(null, l)).sort(sortxy);
```

Comment:
As far as I can see, all calls to `crossxy` involve `l` as one of the arguments. And since `crossxy` is in the same scope as `l`, you could just let that function refer to `l` and a single other argument, and then avoid the binding.